pkg/components/state/pluggable.go

```diff
@@ -35,6 +40,106 @@ var (
 	ErrTransactOperationNotSupported = errors.New("transact operation not supported")
 )
 
+// errors code
+var (
+	GRPCCodeETagMismatch          = codes.FailedPrecondition
+	GRPCCodeETagInvalid           = codes.InvalidArgument
+	GRPCCodeBulkDeleteRowMismatch = codes.Internal
+)
+
+const (
+	// etagField is the field that should be specified on gRPC error response.
+	etagField = "etag"
+	// affectedRowsMetadataKey is the metadata key used to return bulkdelete mismatch errors affected rows.
+	affectedRowsMetadataKey = "affected"
+	// expectedRowsMetadataKey is the metadata key used to return bulkdelete mismatch errors expected rows.
+	expectedRowsMetadataKey = "expected"
+)
+
+// etagErrFromStatus get the etag error from the given gRPC status, if the error is not an etag kind error the return is the original error.
+func etagErrFromStatus(s status.Status) (error, bool) {
+	details := s.Details()
+	if len(details) != 1 {
```

Comment:
I assume it's intentional that an "ETag error" would only ever have a single details, with a single field violation as opposed to being potentially mixed with other kinds of invalid arguments.  That likely makes sense, as I'm not sure how the runtime would coalesce such errors into a single, semantically clear error, or respond to it.

Reply:
I think it could go either way, just depending on how we want to classify ETag errors. If an ETag error is just a more specific type of "invalid argument" error, then I would tend to think it make sense that it could be just one of potentially several invalid argument errors returned as a bundle (for the caller to iterate through).  If an ETag error is considered a higher-level semantic error, considered and returned only after all lower-level errors (like invalid argument errors), then it would make sense to insist that it exist alone.

I'm not familiar enough with Go error handling to know its conventions with respect to bundles of returned errors.  .NET has its `AggregateException` where it's normal to look through and pull out specific kinds of exceptions, whereas with JS/TS, not so much.

Reply:
I think my answer to Tiago might serve for this thread as well: https://github.com/dapr/dapr/pull/5734#discussion_r1084312145

Overall, I feel like we don't have any strong reason to support aggregating up errors, especially that they will end up of being transmitted over http/grpc protocol, so it's going to be serialized and having different treatment depending on the language runtime